src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -40,5 +53,50 @@ public HubConnectionContext(WritableChannel<HubMessage> output, ConnectionContex
         public virtual HubProtocolReaderWriter ProtocolReaderWriter { get; set; }
 
         public virtual WritableChannel<HubMessage> Output => _output;
+
+        public virtual void Abort()
+        {
+            // If we already triggered the token then noop, this isn't thread safe but it's good enough
+            // to avoid spawning a new task in the most common cases
+            if (_connectionAbortedTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+
+            // We fire and forget since this can trigger user code to run
+            Task.Factory.StartNew(_abortedCallback, this);
+        }
+
+        internal void Abort(Exception exception)
+        {
+            AbortException = ExceptionDispatchInfo.Capture(exception);
+            Abort();
+        }
+
+        // Used by the HubEndPoint only
+        internal Task AbortAsync()
+        {
+            Abort();
+            return _abortCompletedTcs.Task;
+        }
+
+        private static void AbortConnection(object state)
+        {
+            var connection = (HubConnectionContext)state;
+            try
+            {
+                connection._connectionAbortedTokenSource.Cancel();
+            }
+            catch
+            {
+                // TODO: Should we log if the cancellation callback fails? This is more preventative to make sure
```

Comment:
"Should we log if?" YES. YES YOU SHOULD LOG IT. ALWAYS LOG IT. LOG ALL THE THINGS. If you're worried about spam, log to `Debug` or `Trace`. But always log it.

An exception from the abort callback should be an error IMO, as the user callback could always capture it and suppress it.

Reply:
> An exception from the abort callback should be an error IMO, as the user callback could always capture it and suppress it.

Yes it's their code they can catch the exceptions.

> "Should we log if?" YES. YES YOU SHOULD LOG IT. ALWAYS LOG IT. LOG ALL THE THINGS. If you're worried about spam, log to Debug or Trace. But always log it.

I think we'll log from the HubEndpoint instead of logging from here.

Reply:
Where we log it is negotiable, as long as we log it ;)